proxy/stat/server.py

```diff
@@ -1,12 +1,97 @@
 from typing import ClassVar
 
 from common.app_data.server import AppDataServer, AppDataApi
+from common.ethereum.hash import EthAddressField
 from common.config.config import Config
+from common.solana.pubkey import SolPubKey
 from common.stat.api import RpcCallData
 from common.stat.metric import StatRegistry, StatSummary, StatGauge
 from common.stat.prometheus import PrometheusServer
 from common.utils.process_pool import ProcessPool
-from .api import STATISTIC_ENDPOINT, TxPoolData, TxFailData, TxDoneData
+from common.utils.pydantic import HexUIntField
+
+from .api import (
+    OpResourceEarnedTokensBalanceData,
+    OpResourceHolderStatusData,
+    OpResourceSpendingTokensBalanceData,
+    STATISTIC_ENDPOINT,
+    TxPoolData,
+    TxFailData,
+    TxDoneData,
+)
+
+
+class OpResourceStatApi(AppDataApi):
+    name: ClassVar[str] = "ProxyStatistic::OpResource"
+
+    def __init__(self, registry: StatRegistry):
+        super().__init__()
+        self._earned_tokens_balance: dict[str, dict[EthAddressField, HexUIntField]] = {}
+        self._earned_tokens_balance_stat = StatGauge(
+            "op_resource_earned_tokens_balance", "Operator earned token balance", registry=registry
+        )
+
+        self._holder_free_cnt: dict[SolPubKey, int] = {}
+        self._holder_used_cnt: dict[SolPubKey, int] = {}
+        self._holder_disabled_cnt: dict[SolPubKey, int] = {}
+
+        self._holder_free_cnt_stat = StatGauge(
+            "op_resource_holder_free", "Operator holder accounts (free)", registry=registry
+        )
+        self._holder_used_cnt_stat = StatGauge(
+            "op_resource_holder_used", "Operator holder accounts (used)", registry=registry
+        )
+        self._holder_disabled_cnt_stat = StatGauge(
+            "op_resource_holder_disabled", "Operator holder accounts (disabled)", registry=registry
+        )
+        self._holder_blocked_cnt_stat = StatGauge(
+            "op_resource_holder_blocked", "Operator holder accounts (blocked)", registry=registry
+        )
+        self._holder_total_cnt_stat = StatGauge(
+            "op_resource_holder_total", "Operator holder accounts (total)", registry=registry
+        )
+
+        self._spending_tokens_balance: dict[SolPubKey, int] = {}
+        self._spending_tokens_balance_stat = StatGauge(
+            "op_resource_spending_tokens_balance", "Operator resource spending token balance", registry=registry
+        )
+
+    @AppDataApi.method(name="commitOpResourceEarnedTokensBalance")
+    def on_op_resource_earned_tokens_balance(self, data: OpResourceEarnedTokensBalanceData) -> None:
+        label = dict(token_name=data.token_name)
+
+        if data.token_name not in self._earned_tokens_balance:
+            self._earned_tokens_balance[data.token_name] = {}
+
+        self._earned_tokens_balance[data.token_name][data.eth_address] = data.balance
+        self._earned_tokens_balance_stat.set(label, sum(self._earned_tokens_balance[data.token_name].values()))
```

Comment:
An operator should see the full picture:
- detailed statistic for each token on each key
- the total numbers for each token

Reply:
Yep, makes sense now. Will do.